### PR TITLE
Fetch all entries with logcli if limit==0.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
So far logcli enforces a limit for fetching logs. However, sometimes users want to fetch all entries in a time range. This change allows to disable the limit by setting it to `0`. The entries are still fetched in batches. However, the batch loop is only stopped once no more results are returned.

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/logcli/query/query_test.go [hunk 1]
```diff
@@ -460,6 +460,46 @@ func Test_batch(t *testing.T) {
...
+			labelMatcher: "{test=\"simple\"}",
+			forward:      false,
+			// Our batchsize is 2 but each query will also return the overlapping last element from the
+			// previous batch, as such we only get one item per call so we make a lot of calls
+			// Call one:   line10 line9 line8 line7
+			// Call two:   line7 line6b line6a line6
+			// Call three: line6b line6a line6 line5
+			// Call four:  line5 line5 line3 line2
+			// Call five:  line1
+			// Call six:   -
```
Comment (jeschkies) [inline pkg/logcli/query/query_test.go:497]: This is a little annoying. We could break early if the returned results length is less than the batch size. However, the print stream API does not return that length.

**Review by dannykopping — COMMENTED** [review]
Nice! LGTM
I'm concerned though that a user can set the batch size really high and cause excessive memory growth in the query-frontend when very large responses are returned. Is this a valid concern?

**Comment by jeschkies** [comment]
> I'm concerned though that a user can set the batch size really high and cause excessive memory growth in the query-frontend when very large responses are returned. Is this a valid concern?

Yes but they can already do this. However, we are not going to run the queries in parallel. Maybe we should set a maximum for the batch size, aka limit in the Loki API.

**Comment by dannykopping** [comment]
> > I'm concerned though that a user can set the batch size really high and cause excessive memory growth in the query-frontend when very large responses are returned. Is this a valid concern?
> 
> Yes but they can already do this. However, we are not going to run the queries in parallel. Maybe we should set a maximum for the batch size, aka limit in the Loki API.

You're right, and i think we're protected by `max_entries_limit_per_query` as well 👍

**Review by dannykopping — APPROVED** [review]

**Pull request merged**